### PR TITLE
 P1907R1 Inconsistencies with non-type template parameters

### LINKED ISSUE #3401: [2019-11 CWG Motion 8] P1907R1 Inconsistencies with non-type template parameters

### PR DESCRIPTION
**Pull request by zygoloid**
Fixes #3401.

Also fixes NB US 092, US 093, US 100, US 102, US 114 (C++20 CD).

Fixes cplusplus/nbballot#91
Fixes cplusplus/nbballot#92
Fixes cplusplus/nbballot#99
Fixes cplusplus/nbballot#101
Fixes cplusplus/nbballot#113

Fixes cplusplus/papers#653

### CODE REVIEW

**Comment by jensmaurer** [comment]
Augmented initial comment of pull request (not the commit message itself) to link and auto-close the respective NB comment issues.

File: source/templates.tex [hunk 1]
```diff
@@ -367,28 +367,31 @@
...
+\item an lvalue reference type, or
+\item a literal class type with the following properties:
+\begin{itemize}
+\item
+all base classes and non-static data members are public and non-mutable and
+\item
+the types of all bases classes and non-static data members are
+structural types or (possibly multi-dimensional) array thereof.
+\end{itemize}
+\end{itemize}
```
Comment (johelegp) [inline source/templates.tex]: I think this is missing some indentation according to https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#latex-source-code-formatting.

File: source/templates.tex [hunk 2]
```diff
@@ -1898,25 +1892,68 @@
...
+
+\item
+their corresponding non-type \grammarterm{template-argument}{s}
+are template-argument-equivalent (see below)
+after conversion to the type of the \grammarterm{template-parameter}, and
+
+\item
+their corresponding template \grammarterm{template-argument}{s}
+refer to the same template.
+\end{itemize}
```
Comment (johelegp) [inline source/templates.tex]: Items should be indented two spaces.

File: source/templates.tex [hunk 3]
```diff
@@ -1898,25 +1892,68 @@
...
+
+\item
+they are of union type and either
+they both have no active member or
+they have the same active member and their active members are template-argument-equivalent, or
+
+\item
+they are of class type and
+their corresponding direct subobjects and reference members are template-argument-equivalent.
 \end{itemize}
```
Comment (johelegp) [inline source/templates.tex]: Same here. The long items call for two-space indentation.

That said, the separation of `\item`s by blank lines looks pretty nice, so maybe it's okay. Maybe a new entry to https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#latex-source-code-formatting is needed. I think this style is novel, as I don't remember seeing it before.

File: source/templates.tex (hunk 1 above)
Comment (jensmaurer) [reply source/templates.tex]: We're not yet sure what to do with `\item`; see #3127. Thus, the formatting used here is as good as (most) any.

File: source/templates.tex (hunk 3 above)
Comment (johelegp) [reply source/templates.tex]: The format is actually in accordance to the [new guidelines](https://github.com/cplusplus/draft/issues/3127#issuecomment-549577948), they're just not in the wiki yet. Disregard my review.

**Comment by zygoloid** [comment]
> Please increase the feature-test macro `__cpp_nontype_template_args`.

We absolutely should bump some feature-test macro here, but that is neither requested by the wording paper nor an editorial change. (Also, is that the right one? Should we bump `__cpp_nontype_template_parameter_class` instead? Or maybe both?) If SG10 and CWG are happy, I think I could be OK with making the change editorially anyway, but it seems like a stretch.

I've started a thread on the relevant reflectors.

**Comment by jensmaurer** [comment]
We're also editorially bumping __cplusplus, so bumping a feature-test macro seems similarly editorial to me. Having SG10 and CWG buy-in is good, though.

**Comment by jensmaurer** [comment]
So, the actual changes are good, but I continue to believe we should bump a feature-test macro. I agree that removing one (as suggested on the reflector) is a bit beyond editorial.

**Comment by zygoloid** [comment]
@BRevzin Can you take a look and confirm you consider the feature test macro changes to be acceptable as an editorial change?

**Comment by BRevzin** [comment]
@zygoloid Yep. There were no objections on either list. Looks good to me.

**Pull request merged**